ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2835,194 @@ public sealed class Bot : IAsyncDisposable {
 			}
 		}
 
+		private void OnInventoryChanged() {
+			Utilities.InBackground(CardsFarmer.OnNewItemsNotification);
+
+			if (BotConfig.BotBehaviour.HasFlag(BotConfig.EBotBehaviour.DismissInventoryNotifications)) {
+				Utilities.InBackground(ArchiWebHandler.MarkInventory);
+			}
+
+			if (BotConfig.SendSetsOnCompleted) {
+				Utilities.InBackground(SendCompletedSets);
+			}
+		}
+
+		private async Task<HashSet<uint>> GetPossiblyCompletedBadgeAppIDs() {
+			using IDocument? badgePage = await ArchiWebHandler.GetBadgePage(1).ConfigureAwait(false);
+
+			if (badgePage == null) {
+				ArchiLogger.LogGenericWarning(Strings.WarningCouldNotCheckBadges);
+
+				return new HashSet<uint>(0);
+			}
+
+			byte maxPages = 1;
+			IElement? htmlNode = badgePage.SelectSingleNode("(//a[@class='pagelink'])[last()]");
+
+			if (htmlNode != null) {
+				string lastPage = htmlNode.TextContent;
+
+				if (string.IsNullOrEmpty(lastPage)) {
+					ArchiLogger.LogNullError(nameof(lastPage));
+
+					return new HashSet<uint>(0);
+				}
+
+				if (!byte.TryParse(lastPage, out maxPages) || (maxPages == 0)) {
+					ArchiLogger.LogNullError(nameof(maxPages));
+
+					return new HashSet<uint>(0);
+				}
+			}
+
+			HashSet<uint> appIDs = GetPossiblyCompletedBadgeAppIDs(badgePage);
+			byte currentPage = 1;
+
+			// We need to check all badge pages, as a badge which we already crafted to max level might hide somewhere in the back
+			while (currentPage <= maxPages) {
+				appIDs.UnionWith(await GetPossiblyCompletedBadgeAppIDs(++currentPage).ConfigureAwait(false));
```

Comment:
Handle `null` result by aborting the whole operation. Also it's entirely unacceptable that this function calls each page synchronously while all of them could be called at once - use ASF approach with `Utilities.InParallel` for handling all pages `>= 2` at the same time. It might or might not require a concurrent collection, depending how you do it - check `CardsFarmer` for examples.